### PR TITLE
UY-1102 additional features for furms 

### PR DESCRIPTION
**Pull request by bihu1**

### CODE REVIEW

File: engine-api/src/main/java/pl/edu/icm/unity/engine/api/AttributesManagement.java [hunk 1]
```diff
@@ -99,4 +101,9 @@
 	Collection<AttributeExt> getAllAttributes(EntityParam entity, boolean effective, 
 			String groupPath, String attributeTypeId, boolean allowDegrade) throws EngineException;
 
+	/**
+	 * Returns attributes of an entity, including hidden ones.
```
Comment (golbi) [inline engine-api/src/main/java/pl/edu/icm/unity/engine/api/AttributesManagement.java]: The key thing here is that attributes from multiple groups are returned, and groups are specified by patterns.

File: engine-api/src/main/java/pl/edu/icm/unity/engine/api/GroupsManagement.java [hunk 2]
```diff
@@ -32,6 +32,13 @@
 	 */
 	void addGroup(Group toAdd) throws EngineException;
 
+	/**
+	 * Adds a new group in a recursively way
+	 * @param toAdd group to add
+	 * @param recursive flag to add group recursively
```
Comment (golbi) [inline engine-api/src/main/java/pl/edu/icm/unity/engine/api/GroupsManagement.java]: After some thinking I guess "recursive" is not a right term here. This is rather "WithParents".

Also note that the API should not be developed like this: boolean method args are considered generally an anti pattern. Instead you can have two methods, named properly, without a boolean param at all. There are sometimes practical exceptions, but that can be considered a rule of thumb.

File: rest-admin/src/test/java/pl/edu/icm/unity/restadm/TestGroupsManagement.java [hunk 3]
```diff
@@ -60,6 +61,25 @@ public void addedGroupIsReturned() throws Exception
 		assertThat(groupContent.getSubGroups().contains("/subgroup"), is(true));
 	}
 
+	@Test
+	public void recursivelyAddedGroupIsReturned() throws Exception
```
Comment (golbi) [inline rest-admin/src/test/java/pl/edu/icm/unity/restadm/TestGroupsManagement.java:65]: Well, we can have an integration test here (optionally really), but a must have is to test the engine operation. So tests like this are needed in engine. And perhaps not one but at least two (creating all parents, or few parents).

**Review by golbi — CHANGES_REQUESTED** [review]

File: types-api/src/main/java/pl/edu/icm/unity/types/basic/GroupPattern.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2015 ICM Uniwersytet Warszawski All rights reserved.
+ * See LICENCE.txt file for licensing information.
+ */
+package pl.edu.icm.unity.types.basic;
+
+public class GroupPattern {
```
[the remark this answers is not in the document]
Comment (bihu1) [reply types-api/src/main/java/pl/edu/icm/unity/types/basic/GroupPattern.java]: done

File: documentation/src/main/rest-api/rest-api-v1.txt [hunk 5]
```diff
@@ -196,6 +196,76 @@ Example output:
...
 
+=== Get entity attributes in groups
+
++@Path(/entity/{entityId}/groups/attributes)+ +
++@QueryParam("groupsPatterns")+ +
++@QueryParam("effective")+ +
++@QueryParam("identityType")+ +
++@GET+ +
+
+Returns grouped by group attributes of a given entity in a selected groups patterns.
```
[the remark this answers is not in the document]
Comment (bihu1) [reply documentation/src/main/rest-api/rest-api-v1.txt:207]: done

File: engine-api/src/main/java/pl/edu/icm/unity/engine/api/GroupsManagement.java [hunk 6]
```diff
@@ -32,6 +32,13 @@
 	 */
 	void addGroup(Group toAdd) throws EngineException;
 
+	/**
+	 * Adds a new group in a recursively way
```
[the remark this answers is not in the document]
Comment (bihu1) [reply engine-api/src/main/java/pl/edu/icm/unity/engine/api/GroupsManagement.java]: done

File: engine-api/src/main/java/pl/edu/icm/unity/engine/api/GroupsManagement.java (hunk 2 above)
Comment (bihu1) [reply engine-api/src/main/java/pl/edu/icm/unity/engine/api/GroupsManagement.java]: done

File: documentation/src/main/rest-api/rest-api-v1.txt [hunk 7]
```diff
@@ -334,10 +404,13 @@ Output structure:
 === Create group (simple)
 
 +@Path("/group/{groupPath}")+ +
++@QueryParam("recursive")+ +
 +@POST+ +
 
 Creates a new group. The created group will be empty and will contain only the default settings 
 (e.g. no attribute statements, no description, displayed name set to group path name).
+The optional recursive query parameter (by default false) can be used to add groups in recursively way.
```
[the remark this answers is not in the document]
Comment (bihu1) [reply documentation/src/main/rest-api/rest-api-v1.txt]: done

File: engine-api/src/main/java/pl/edu/icm/unity/engine/api/AttributesManagement.java (hunk 1 above)
Comment (bihu1) [reply engine-api/src/main/java/pl/edu/icm/unity/engine/api/AttributesManagement.java]: done

File: rest-admin/src/test/java/pl/edu/icm/unity/restadm/TestGroupsManagement.java (hunk 3 above)
Comment (bihu1) [reply rest-admin/src/test/java/pl/edu/icm/unity/restadm/TestGroupsManagement.java:65]: done

File: engine/src/test/java/pl/edu/icm/unity/engine/attribute/AttributesManagementImplTest.java [hunk 8]
```diff
@@ -308,11 +310,35 @@ public void updatedAttributeIsReturned() throws Exception
...
 		assertEquals("333", getAttributeByName(allAts, "tel").getValues().get(0));
 		
-		allAts = attrsMan.getAllAttributes(entity, true, null, null, false);
+		allAts = attrsMan.getAllAttributes(entity, true, (String)null, null, false);
 		assertEquals(DEF_ATTRS + 1, allAts.size());
 		assertEquals("333", getAttributeByName(allAts, "tel").getValues().get(0));
 	}
 
+	@Test
+	public void updatedAttributeIsReturnedByGroupPattern() throws Exception
```
[the remark this answers is not in the document]
Comment (bihu1) [reply engine/src/test/java/pl/edu/icm/unity/engine/attribute/AttributesManagementImplTest.java]: done

**Comment by golbi** [comment]
I'm merging that now - much better. There are two minor things missed:
1. no info in rest api doc how to encode multiple group patterns (string with items separated by comma AFAIR)
2. perhaps we have small discussion to be done about boolean params - but let's leave it as-is, not a big deal.

**Pull request merged**